src/scales/scale.time.js

```diff
@@ -361,27 +361,54 @@ function generate(min, max, capacity, options) {
  * Returns the right and left offsets from edges in the form of {left, right}.
  * Offsets are added when the `offset` option is true.
  */
-function computeOffsets(table, ticks, min, max, options) {
+function computeOffsets(table, ticks, data, min, max, options) {
+	var pos = [];
+	var minInterval = 1;
+	var timeOpts = options.time;
+	var barThickness = options.barThickness;
 	var left = 0;
 	var right = 0;
-	var upper, lower;
-
-	if (options.offset && ticks.length) {
-		if (!options.time.min) {
-			upper = ticks.length > 1 ? ticks[1] : max;
-			lower = ticks[0];
-			left = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
+	var i, ilen, curr, prev, length, width;
+
+	if (options.offset) {
+		data.forEach(function(timestamp) {
+			if (timestamp >= min && timestamp <= max) {
+				pos.push(interpolate(table, 'time', timestamp, 'pos'));
+			}
+		});
+
+		if (!barThickness) {
+			[data, ticks].forEach(function(timestamps) {
+				for (i = 0, ilen = timestamps.length; i < ilen; ++i) {
+					curr = interpolate(table, 'time', timestamps[i], 'pos');
+					minInterval = i > 0 ? Math.min(minInterval, curr - prev) : minInterval;
+					prev = curr;
+				}
+			});
 		}
-		if (!options.time.max) {
-			upper = ticks[ticks.length - 1];
-			lower = ticks.length > 1 ? ticks[ticks.length - 2] : min;
-			right = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
+
+		length = pos.length;
+		if (length) {
```

Comment:
Can this be combined with the enclosing `if` statement? I.e. `if (options.offset && data.length)`?

Reply:
Even if `data` is not empty, we need to use the subset of `data`, which only contains timestamps between `min` and `max`, so checking `pos.length` is needed. But, we can move this `if (length)` statement before the calculation of `minInterval`.

Reply:
Ah, you're right. Maybe we could also invert the check. E.g.:

```
		data.forEach(function(timestamp) {
			if (timestamp >= min && timestamp <= max) {
				pos.push(interpolate(table, 'time', timestamp, 'pos'));
			}
		});

		length = pos.length;
		if (!length) {
			return;
		}
```

Reply:
As this function needs to return an object, I will just move the check.